assets/tests/hooks/useVehiclesSocket.test.ts

```diff
@@ -0,0 +1,134 @@
+import { Channel, Socket } from "phoenix"
+import { renderHook } from "react-hooks-testing-library"
+import useVehiclesSocket, {
+  initialState,
+  reducer,
+  removeRoute,
+  setChannelForRoute,
+  setSocket,
+  setVehiclesForRoute,
+} from "../../src/hooks/useVehiclesSocket"
+import { RouteId, Vehicle } from "../../src/skate"
+
+// tslint:disable: react-hooks-nesting
+
+const mockSocketChannel = jest.fn()
+const mockSocketConnect = jest.fn()
+
+jest.mock("phoenix", () => ({
+  Socket: jest.fn(() => ({
+    channel: mockSocketChannel,
+    connect: mockSocketConnect,
+  })),
+  __esModule: true,
+}))
+
+const makeMockChannel = () => {
+  const result = {
+    join: jest.fn(),
+    leave: jest.fn(),
+    on: jest.fn(),
+    receive: jest.fn(),
+  }
+  result.join.mockImplementation(() => result)
+  result.receive.mockImplementation(() => result)
+  return result
+}
+
+describe("useVehiclesSocket", () => {
+  test("vehicles is empty to start with", () => {
+    const { result } = renderHook(() => useVehiclesSocket([]))
+
+    expect(mockSocketConnect).toHaveBeenCalled()
+    expect(result.current).toEqual({})
+  })
+
+  test("selecting a new route subscribes to the new channel", () => {
+    const mockChannel = makeMockChannel()
+    mockSocketChannel.mockImplementationOnce(() => mockChannel)
+
+    const { rerender } = renderHook(() => useVehiclesSocket(["1"]))
+
+    // Needs to be kicked to do the effects again after the socket initializes
+    rerender()
+
+    expect(mockSocketChannel).toHaveBeenCalledTimes(1)
+    expect(mockSocketChannel).toHaveBeenCalledWith("vehicles:1")
+    expect(mockChannel.join).toHaveBeenCalledTimes(1)
+  })
+
+  test("unselecting a route unsubscribes from the channel", async () => {
+    const mockChannel = makeMockChannel()
+    mockSocketChannel.mockImplementationOnce(() => mockChannel)
+
+    const { rerender } = renderHook(
+      (selectedRouteIds: RouteId[]) => useVehiclesSocket(selectedRouteIds),
+      { initialProps: ["1"] }
+    )
+    rerender([]) // Deselect the route
+
+    expect(mockChannel.leave).toHaveBeenCalledTimes(1)
+  })
+})
+
+describe("reducer", () => {
```

Comment:
Is this test needed?

It's peeking pretty deeply at implementation details (as opposed to the `useVehiclesSocket` test, which doesn't need to know anything about how the hook stores its internal state). If we got rid of it, we could get rid of all the exports from `useVehiclesSocket.ts` except for the hook itself, which would be nice. But until we figure out how to test data arriving, this tests some code that the more blackbox testing can't.

Reply:
I managed to test the return results (by synchronously calling the handler passed into `.on()` and `.receive()`), so testing the reducer implementation is no longer necessary. Removed this.